### PR TITLE
Use DPP runner instead of API loader

### PR DESCRIPTION
**Pull request by akariv**
DPP Runner (https://github.com/akariv/dpp-runner) is a small utility library for running a single pipeline/source spec in a temporary workspace and as a separate thread.
You can submit these specs to it, and it will call a callback function with status updates.

The modifications in this PR:
- [ ] Remove the status callback from os conductor (the one that API loader used to call), and repurposes the code to serve as the dpp-runner callback.
- [ ] Modifies the package/upload handler, to create a fiscal source-spec from the provided datapackage and use that spec as an input to dpp-runner
- [ ] Modifications to the Dockerfile to allow all that goodness to actually work in a single container.

(there are still a couple of tests failing, will take care of them)

### CODE REVIEW

File: conductor/blueprints/package/blueprint.py [hunk 1]
```diff
@@ -35,8 +35,7 @@ def upload():
         abort(400)
     if jwt is None:
         abort(403)
-    callback = os_conductor+url_for('.callback')
```
Comment (brew) [inline conductor/blueprints/package/blueprint.py]: I think this means you can get rid of the `os_conductor` variable on L16, and the `OS_CONDUCTOR_URL` env var completely (and its mention in the README), as it's only there to help define the callback url.

File: conductor/blueprints/package/controllers.py [hunk 2]
```diff
@@ -32,25 +75,36 @@ def upload(datapackage, callback, token, cache_set):
...
     else:
-        params = {
-            'package': datapackage,
-            'callback': callback
+        key = 'os-conductor:package:'+datapackage
+        ret = {
+            "progress": 0,
+            "status": "queued"
         }
-        load_url = os_api_url+'/api/3/loader/'
```
Comment (brew) [inline conductor/blueprints/package/controllers.py]: Does this mean we no longer need the os-api-loader service (the os-api app configured as a loader)?

File: conductor/blueprints/package/controllers.py [hunk 3]
```diff
@@ -60,24 +114,49 @@ def upload_status(datapackage, cache_get):
...
+        if 'INPROGRESS' in statuses:
+            return 'loading-data'
+        if all(self.statuses.get(pi) == 'SUCCESS' 
+               for pi in ('./finalize_datapackage_flow', './dumper_flow_update_status')):
+            return 'done'
+        return 'loading-data'
+
+
+    def __call__(self, pipeline_id, status, errors=None, stats=None):
+        logging.error('upload_status_update: %s pipeline:%s, ' +
```
Comment (brew) [inline conductor/blueprints/package/controllers.py]: logging.error?

File: requirements.txt [hunk 4]
```diff
@@ -5,14 +5,18 @@ flask-jsonpify
...
 pyjwt
 sqlalchemy
 psycopg2
 cryptography
 elasticsearch>=1.0.0,<2.0.0
-os-package-registry>=0.0.12
+os-package-registry>=0.0.13
+datapackage-pipelines-fiscal>=1.0.7
+datapackage-pipelines-aws
+datapackage-pipelines[speedup]
```
Comment (brew) [inline requirements.txt]: I've not used that square-brackets syntax before. Learned something new!

File: Dockerfile [hunk 5]
```diff
@@ -1,23 +1,16 @@
...
+RUN apk add --update --no-cache --virtual=build-dependencies build-base 
+RUN apk --repository http://dl-3.alpinelinux.org/alpine/edge/testing/ --update add leveldb leveldb-dev
 
 WORKDIR /app
 ADD requirements.txt .
 RUN pip install -r requirements.txt
+RUN npm install -g os-types
+
+RUN apk del build-dependencies
+RUN rm -rf /var/cache/apk/*
```
Comment (brew) [inline Dockerfile]: Can L12-13 be put before `WORKDIR /app` etc, as the results will change less frequently than requirement.txt installation, making better use of the Docker cache and help speed up Docker builds.

File: Dockerfile [hunk 6]
```diff
@@ -1,23 +1,16 @@
...
-    libpq \
-    wget \
-    libffi \
-    libffi-dev \
-    ca-certificates \
-    python3-dev \
-    postgresql-dev \
-    build-base \
-    bash \
-    curl
```
Comment (brew) [inline Dockerfile]: We need bash and curl to make the [`entrypoint.sh`](https://github.com/openspending/os-conductor/blob/198bf5c0223818422a053453566545afd56c30cb/docker/docker-entrypoint.sh) script work.

File: Dockerfile [hunk 7]
```diff
@@ -1,23 +1,16 @@
...
-RUN update-ca-certificates
+RUN apk add --update --no-cache libpq postgresql-dev libffi libffi-dev bash libstdc++ nodejs
+RUN apk add --update --no-cache --virtual=build-dependencies build-base 
+RUN apk --repository http://dl-3.alpinelinux.org/alpine/edge/testing/ --update add leveldb leveldb-dev
 
 WORKDIR /app
 ADD requirements.txt .
 RUN pip install -r requirements.txt
+RUN npm install -g os-types
```
Comment (brew) [inline Dockerfile]: What's this used for?

File: conductor/blueprints/package/blueprint.py (hunk 1 above)
Comment (akariv) [reply conductor/blueprints/package/blueprint.py]: will do

File: conductor/blueprints/package/controllers.py (hunk 2 above)
Comment (akariv) [reply conductor/blueprints/package/controllers.py]: indeed. The entire service can be removed from os-api

File: conductor/blueprints/package/controllers.py (hunk 3 above)
Comment (akariv) [reply conductor/blueprints/package/controllers.py]: yes, debugging :) will remove

File: Dockerfile (hunk 5 above)
Comment (akariv) [reply Dockerfile]: build-dependencies is needed for the pip install - we install it for installing python libs with native code.

File: Dockerfile (hunk 7 above)
Comment (akariv) [reply Dockerfile]: This is the code that builds the FDP model - it's used in the packager as well.

File: Dockerfile (hunk 6 above)
Comment (akariv) [reply Dockerfile]: okay

File: conductor/blueprints/package/controllers.py (hunk 2 above)
Comment (pwalsh) [reply conductor/blueprints/package/controllers.py]: @brew @akariv should we perhaps merge os-api and os-conductor into a single codebase?

File: Dockerfile (hunk 7 above)
Comment (brew) [reply Dockerfile]: Is it used in the os-conductor codebase, I can't find it.

File: Dockerfile (hunk 5 above)
Comment (brew) [reply Dockerfile]: Sure, but can these lines me moved before the `WORKDIR /app` line. It just optimises the caching.

File: Dockerfile (hunk 7 above)
Comment (akariv) [reply Dockerfile]: It is used by datapackage-pipelines-fiscal, inside the model.py processor

File: Dockerfile (hunk 5 above)
Comment (akariv) [reply Dockerfile]: No - they must come after `pip install -r requirements.txt`, which is after `WORKDIR /app`

They are before the `ADD . .` which is what mostly changes.

Comment (brew) [reply Dockerfile]: Gotcha - I'd misunderstood what `del` was doing.

File: Dockerfile (hunk 7 above)
Comment (brew) [reply Dockerfile]: Can you add a comment here to say what's using it, and a note in datapackage-pipelines-fiscal's README file to state it as a dependency.

File: conductor/blueprints/package/blueprint.py (hunk 1 above)
Comment (brew) [reply conductor/blueprints/package/blueprint.py]: And `OS_API_URL` too :)

File: Dockerfile (hunk 7 above)
Comment (akariv) [reply Dockerfile]: will do.

File: conductor/blueprints/package/controllers.py (hunk 2 above)
Comment (akariv) [reply conductor/blueprints/package/controllers.py]: With this change these services are completely disconnected - 
so it became quite simple to do it (while there's no special gain from doing so)

File: conductor/blueprints/package/blueprint.py (hunk 1 above)
Comment (akariv) [reply conductor/blueprints/package/blueprint.py]: done

File: conductor/blueprints/package/controllers.py (hunk 2 above)
Comment (akariv) [reply conductor/blueprints/package/controllers.py]: Removed all loader functionality in the corresponding os-api pull request.

File: conductor/blueprints/package/controllers.py (hunk 3 above)
Comment (akariv) [reply conductor/blueprints/package/controllers.py]: changed to logging.debug

File: requirements.txt (hunk 4 above)
Comment (akariv) [reply requirements.txt]: 👍

File: Dockerfile (hunk 6 above)
Comment (akariv) [reply Dockerfile]: bash and curl - check!

File: Dockerfile (hunk 7 above)
Comment (akariv) [reply Dockerfile]: done + https://github.com/openspending/datapackage-pipelines-fiscal/commit/64bf315e53cec4fdf8f0a03737f5f63fd38f8fa4

**Comment by akariv** [comment]
@brew fixed CR, tests and Docker.

File: .travis.yml [hunk 8]
```diff
@@ -25,6 +25,7 @@ before_install:
 
 install:
   # Will build local Dockerfile as part of ci-run (docker-compose up)
+  - sudo apt-get install libleveldb-dev libleveldb1
```
Comment (brew) [inline .travis.yml]: What's this used by? Do we need it in the Dockerfile instead, or is it just for Travis?

**Comment by brew** [comment]
Nice! A joy to get rid of more env vars :D

File: .travis.yml (hunk 8 above)
Comment (akariv) [reply .travis.yml]: It's actually already in the dockerfile - these are native libraries used by the `speedup` extension for `dpp`.

File: Dockerfile [hunk 9]
```diff
@@ -1,24 +1,19 @@
...
-    ca-certificates \
-    python3-dev \
-    postgresql-dev \
-    build-base \
-    bash \
-    curl
-RUN update-ca-certificates
+RUN apk add --update --no-cache libpq postgresql-dev libffi libffi-dev bash curl libstdc++ nodejs
+RUN apk add --update --no-cache --virtual=build-dependencies build-base 
+RUN apk --repository http://dl-3.alpinelinux.org/alpine/edge/testing/ --update add leveldb leveldb-dev
```
Comment (akariv) [inline Dockerfile]: @brew here ^

**Pull request merged**